File: 0236-lowest-common-ancestor-of-a-binary-tree-二叉树的最近公共祖先/0236.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        stack = [[root]]
        ancestorsP = []
        ancestorsQ = []
        while stack:
            nodes = stack.pop()
            if not nodes[-1]:
                continue
            if len(ancestorsP) > 0 and len(ancestorsQ) > 0:
                # 若双方的祖先节点都找到, 则终止
                break
            if nodes[-1] == p:
                ancestorsP = nodes
            if nodes[-1] == q:
                ancestorsQ = nodes
            left = nodes.copy()
            left.append(nodes[-1].left)
            right = nodes.copy()
            right.append(nodes[-1].right)
            stack.append(left)
            stack.append(right)


        # 寻找双方祖先节点的交点, 解法类似第160题相交链表
        aP = ancestorsQ + ancestorsP
        aQ = ancestorsP + ancestorsQ
        pointerP = len(aP) - 1
        pointerQ = len(aQ) - 1
        while aP[pointerP] != aQ[pointerQ]:
            pointerP -= 1
            pointerQ -= 1
        return aP[pointerP]
            

            

File: 0236-lowest-common-ancestor-of-a-binary-tree-二叉树的最近公共祖先/0236.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    vector<pair<TreeNode*, int>> paths;
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        int i = -1;
        TreeNode* nodeFound = NULL;
        paths.push_back(make_pair(root, 0));

        while (true) {
            auto &nodePair = paths.back();
            if (!nodePair.first) {
                paths.pop_back();
                continue;
            };
            // cout << nodePair.first->val << "," << i << endl;

            if (nodePair.second != 2) {
                if (nodePair.first == p || nodePair.first == q) {
                    if (i == -1) {
                        i = paths.size() - 1;
                        nodeFound = nodePair.first;
                    } else if (nodePair.first != nodeFound) return paths[i].first;
                }
                if (nodePair.second == 0) {
                    nodePair.second = 1;
                    paths.push_back(make_pair(nodePair.first->left, 0));
                } else if (nodePair.second == 1) {
                    nodePair.second = 2;
                    paths.push_back(make_pair(nodePair.first->right, 0));
                }
                
            } else {
                paths.pop_back();
                int size = paths.size();
                if (i >= size) {
                    i--;
                }
            }
        }
        return NULL;
    }
};